File: Data Structures & Algorithms/construct-quad-tree/submission-0.java
/*
// Definition for a QuadTree node.
class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;


    public Node() {
        this.val = false;
        this.isLeaf = false;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }

    public Node(boolean val, boolean isLeaf) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }

    public Node(boolean val, boolean isLeaf, Node topLeft, Node topRight, Node bottomLeft, Node
bottomRight) { this.val = val; this.isLeaf = isLeaf; this.topLeft = topLeft; this.topRight =
topRight; this.bottomLeft = bottomLeft; this.bottomRight = bottomRight;
    }
}
*/

class Solution {
    public boolean isSame(int[][] g, int x, int y, int size) {
        int val = g[x][y];
        for (int i = x; i < x + size; i++) {
            for (int j = y; j < y + size; j++) {
                if (g[i][j] != val) {
                    return false;
                }
            }
        }
        return true;
    }

    public Node dfs(int[][] g, int x, int y, int size) {
        if (isSame(g, x, y, size)) {
            return new Node(g[x][y] == 1, true);
        }

        int half = size / 2;
        Node root = new Node(true, false);

        root.topLeft = dfs(g, x, y, half);
        root.topRight = dfs(g, x, y + half, half);
        root.bottomLeft = dfs(g, x + half, y, half);
        root.bottomRight = dfs(g, x + half, y + half, half);

        return root;
    }

    public Node construct(int[][] grid) {
        return dfs(grid, 0, 0, grid.length);
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-0.cpp
class Solution {
   public:
    int dfs(vector<vector<int>>& g, int r, int c) {
        if (r < 0 || c < 0 || r >= g.size() || c >= g[0].size() || g[r][c] == 0) {
            return 0;
        }

        g[r][c] = 0;
        int res = 1;

        res += dfs(g, r + 1, c);
        res += dfs(g, r - 1, c);
        res += dfs(g, r, c + 1);
        res += dfs(g, r, c - 1);

        return res;
    }

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int r = grid.size();
        int c = grid[0].size();
        int area = 0;

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (grid[i][j] == 1) {
                    area = max(area, dfs(grid, i, j));
                }
            }
        }
        return area;
    }
};